lib/guardian/topic_guardian.rb

```diff
@@ -199,7 +199,80 @@ def can_see_deleted_topics?(category)
     is_staff? || is_category_group_moderator?(category)
   end
 
+  # Accepts an array of `Topic#id` and returns an array of `Topic#id` which the user can see.
+  def can_see_topic_ids(topic_ids: [], hide_deleted: true)
+    return topic_ids if is_admin?
+    return [] if topic_ids.compact.blank?
+
+    scope = Topic.unscoped.where(id: topic_ids)
+
+    # When `hide_deleted` is `true`, hide deleted topics if user is not staff or category moderator
+    if hide_deleted && !is_staff?
+      scope = scope = scope.where("deleted_at IS NULL")
+
+      if authenticated?
+        group_moderator_category_ids_sql = Category
+          .joins(reviewable_by_group: :group_users)
+          .where("group_users.user_id = ?", user.id)
+          .select(:id)
+          .to_sql
+
+        scope = scope.or(
+          Topic.unscoped.where("deleted_at IS NOT NULL AND topics.category_id IN (#{group_moderator_category_ids_sql})")
+        )
```

Comment:
Leaving a note before I leave for the weekend but this is duplicating the `is_category_moderator` logic. This introduces the risk where the logic in that method may differ from the query here.